### ISSUE TITLE
Stack overflow for a recursive type constraint

### ISSUE THREAD

**Issue by lawrencejones** [dominikh/go-tools#1262]
Please make sure to include the following information in your issue report:

- The output of 'staticcheck -version'
```
$ staticcheck -version
staticcheck 2022.1.1 (v0.3.1)
```
- The output of 'staticcheck -debug.version' (it is fine if this command fails)
```
$ staticcheck -debug.version
staticcheck 2022.1.1 (v0.3.1)

Compiled with Go version: go1.18
Main module:
        honnef.co/go/tools@v0.3.1 (sum: h1:1kJlrWJLkaGXgcaeosRXViwviqjI7nkBvU2+sZW0AYc=)
Dependencies:
        github.com/BurntSushi/toml@v1.1.0 (sum: h1:ksErzDEI1khOiGPgpwuI7x2ebx/uXQNw7xJpn9Eq1+I=)
        golang.org/x/exp/typeparams@v0.0.0-20220426173459-3bcf042a4bf5 (sum: h1:pKfHvPtBtqS0+V/V9Y0cZQa2h8HJV/qSRJiGgYu+LQA=)
        golang.org/x/mod@v0.6.0-dev.0.20220106191415-9b9b3d81d5e3 (sum: h1:kQgndtyPBW/JIYERgdxfwMYh3AVStj88WQTlNDi2a+o=)
        golang.org/x/sys@v0.0.0-20220422013727-9388b58f7150 (sum: h1:xHms4gcpe1YE7A3yIllJXP16CMAGuqwO2lX1mTyyRRc=)
        golang.org/x/tools@v0.1.11-0.20220316014157-77aa08bb151a (sum: h1:ofrrl6c6NG5/IOSx/R1cyiQxxjqlur0h/TvbUhkH0II=)
        golang.org/x/xerrors@v0.0.0-20220411194840-2f41105eb62f (sum: h1:GGU+dLjvlC3qDwqYgL6UgRmHXhOOgns0bZu2Ty5mm6U=)
```
- The output of 'go version'
```
go version
go version go1.18 darwin/arm64
```
- The output of 'go env'
```
go env
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/lawrence/Library/Caches/go-build"
GOENV="/Users/lawrence/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/lawrence/Projects/golang/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/lawrence/Projects/golang"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/opt/homebrew/Cellar/go/1.18/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/opt/homebrew/Cellar/go/1.18/libexec/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.18"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/lawrence/Projects/core/server/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/r7/sjj514qn2vqf5kc9x_vq9s480000gn/T/go-build3330339786=/tmp/go-build -gno-record-gcc-switches -fno-common"
```
- Exactly which command you ran
```
$ staticcheck github.com/incident-io/core/server/pkg/service/event
```
- Output of the command and what's wrong with the output
```
runtime: goroutine stack exceeds 1000000000-byte limit
runtime: sp=0x1402d3803b0 stack=[0x1402d380000, 0x1404d380000]
fatal error: stack overflow

runtime stack:
runtime.throw({0x102f6be26?, 0x103338140?})
        /opt/homebrew/Cellar/go/1.18/libexec/src/runtime/panic.go:992 +0x50
runtime.newstack()
        /opt/homebrew/Cellar/go/1.18/libexec/src/runtime/stack.go:1101 +0x46c
runtime.morestack()
        /opt/homebrew/Cellar/go/1.18/libexec/src/runtime/asm_arm64.s:310 +0x70

goroutine 726 [running]:
golang.org/x/tools/go/types/objectpath.find({0x1030f3578?, 0x140016ff7c0?}, {0x1030ed348?, 0x14001d8f500?}, {0x1400a78a000, 0x2811b8, 0x2a0000})
        /Users/lawrence/Projects/golang/pkg/mod/golang.org/x/tools@v0.1.11-0.20220316014157-77aa08bb151a/go/types/objectpath/objectpath.go:319 +0xd1c fp=0x1402d3803b0 sp=0x1402d3803b0 pc=0x102e7c29c
golang.org/x/tools/go/types/objectpath.find({0x1030f3578?, 0x140016ff7c0?}, {0x1030ed4b0?, 0x14001d8df20?}, {0x1400a78a000, 0x2811b7, 0x2a0000})
        /Users/lawrence/Projects/golang/pkg/mod/golang.org/x/tools@v0.1.11-0.20220316014157-77aa08bb151a/go/types/objectpath/objectpath.go:387 +0x40c fp=0x1402d380520 sp=0x1402d3803b0 pc=0x102e7b98c
...additional frames elided...
created by honnef.co/go/tools/lintcmd/runner.(*Runner).Run
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/runner/runner.go:1307 +0x4dc

goroutine 1 [chan send]:
honnef.co/go/tools/internal/sync.Semaphore.Acquire(...)
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/internal/sync/sync.go:14
honnef.co/go/tools/lintcmd/runner.(*Runner).Run(0x14000274820, 0x102cac2fc?, {0x14000103680?, 0x102b90780?, 0x1?}, {0x14000020050, 0x1, 0x1})
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/runner/runner.go:1306 +0x510
honnef.co/go/tools/lintcmd.(*linter).Lint(0x1400023a9f0, 0x10335ea70?, {0x14000020050?, 0x1, 0x1})
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/lint.go:93 +0x1b4
honnef.co/go/tools/lintcmd.doLint({0x14000103200, 0x90, 0x129ffef00?}, {0x14000020050, 0x1, 0x1}, 0x129ffee00?)
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/lint.go:572 +0x480
honnef.co/go/tools/lintcmd.(*Command).Run(0x14000196360)
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/cmd.go:422 +0xfe4
main.main()
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/cmd/staticcheck/staticcheck.go:44 +0x518

goroutine 87 [syscall]:
os/signal.signal_recv()
        /opt/homebrew/Cellar/go/1.18/libexec/src/runtime/sigqueue.go:148 +0x30
os/signal.loop()
        /opt/homebrew/Cellar/go/1.18/libexec/src/os/signal/signal_unix.go:23 +0x20
created by os/signal.Notify.func1.1
        /opt/homebrew/Cellar/go/1.18/libexec/src/os/signal/signal.go:151 +0x30

goroutine 88 [chan receive]:
honnef.co/go/tools/lintcmd.doLint.func2()
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/lint.go:567 +0x50
created by honnef.co/go/tools/lintcmd.doLint
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/lint.go:566 +0x464

goroutine 716 [chan send]:
honnef.co/go/tools/lintcmd/runner.genericHandle({0x1030efcc0?, 0x1400044d320}, {0x1030efcc0?, 0x14000349200?}, 0x0?, 0x140002748e0, 0x14008aed9b0)
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/runner/runner.go:868 +0x1a0
created by honnef.co/go/tools/lintcmd/runner.(*Runner).Run
        /Users/lawrence/Projects/golang/pkg/mod/honnef.co/go/tools@v0.3.1/lintcmd/runner/runner.go:1307 +0x4dc
```
- Where we can read the code you're running Staticcheck on

I haven't got a reproducible snippet right now as when I move the problematic code into a blank package, things work again. I think this is due to staticcheck assuming it's a no-op in a blank package and not trying to parse the code rather than it being anything to do with the rest of the code, but the problematic function is:

```go
func ExpectPublished[T interface {
	Matcher(*T) types.GomegaMatcher
}](opts func(*T)) {
}
```

That is the function that triggers the stack overflow, presumably because the linter is parsing the type constraint which in turn references T, the type constraint.

---

I've attached all the details about what triggers the stack overflow. It looks like we're infinitely recursing on a recursive type constraint.

When I get more time I'll create a minimal repro, but opening this for now in case it's enough to get going (it may well be!).

**Comment by dominikh** [comment]
tl;dr: Please check if Staticcheck at `master` (specifically 84471109f0e246b22c64606755861f439075b2d8 or newer) behaves correctly.

---

If `github.com/incident-io/core` had been public, then I would've reduced it to a minimal sample myself. This way, there's not much I can do without an actual reproducer.

The actual stack overflow occurs in `golang.org/x/tools/go/types/objectpath.find`, which isn't our code. It's possible that this has already been fixed upstream and we need to update our dependency on `golang.org/x/tools`. https://go-review.googlesource.com/c/tools/+/393635/ looks like a likely candidate. I've pushed a commit to master that updates our dependency. Please check if that helps.

Note that

>  I think this is due to staticcheck assuming it's a no-op in a blank package and not trying to parse the code rather than it being anything to do with the rest of the code

isn't really an accurate description, as Staticcheck always does at least some work on the code. Surrounding code may, however, cause it to do more work.

**Comment by lawrencejones** [comment]
Managed to upgrade to master today, as per your suggestion, and very happy to say this is now fixed!

Thank you for your response, really appreciate the help.

Going to close this now. Have a great day :)
